### ISSUE TITLE
AMP Cache: Cached img smaller than AMP origin img

### ISSUE THREAD

**Issue by jamesozzie** [ampproject/amphtml#23453]
AMP Cache optimisations result in smaller image dimensions compared from AMP origin. While it renders different image sizes per screen the initial load image is much smaller than AMP origin for the same size container. 

Standard/Origin AMP: [https://www.reviewcenter.in/8837/lg-dual-inverter-air-conditioner-ac-review-features-specifications/?amp](https://www.reviewcenter.in/8837/lg-dual-inverter-air-conditioner-ac-review-features-specifications/?amp)

Cached AMP site: [https://www-reviewcenter-in.cdn.ampproject.org/c/s/www.reviewcenter.in/8837/lg-dual-inverter-air-conditioner-ac-review-features-specifications/?amp](https://www-reviewcenter-in.cdn.ampproject.org/c/s/www.reviewcenter.in/8837/lg-dual-inverter-air-conditioner-ac-review-features-specifications/?amp)

Additional Examples
[AMP cache](https://www.google.com/amp/s/www.theindianwire.com/news/ayodhya-case-final-decision-november-17-supreme-court-begins-discussion-mediation-panel-204991/%3famp) | [AMP Origin ](https://www.theindianwire.com/news/ayodhya-case-final-decision-november-17-supreme-court-begins-discussion-mediation-panel-204991/?amp)

[AMP Cache](https://www.google.com/amp/s/www.gizmochina.com/2019/09/30/flipkart-big-billion-sale-grab-unmissable-offers-on-smartphones-and-other-electronic-devices/%3famp) | [Origin URL](https://www.gizmochina.com/2019/09/30/flipkart-big-billion-sale-grab-unmissable-offers-on-smartphones-and-other-electronic-devices/?amp)

**Additional Information:**
CMS: WordPress, using the official AMP plugin
Theme: Newspaper by [TagDiv](https://tagdiv.com/amp-newspaper-theme/?amp)
After inspecting other websites using this theme the same issue occurs. 

**AMP origin markup:**
`<img decoding="async" alt="Review Center India" title="Review Center India" src="https://www.reviewcenter.in/wp-content/uploads/2015/05/xreviewcenter.jpg" class="i-amphtml-fill-content i-amphtml-replaced-content">`

**AMP cache markup:**
`<img decoding="async" alt="Review Center India" title="Review Center India" srcset="https://www-reviewcenter-in.cdn.ampproject.org/ii/w270/s/www.reviewcenter.in/wp-content/uploads/2015/05/xreviewcenter.jpg 270w, https://www-reviewcenter-in.cdn.ampproject.org/ii/w560/s/www.reviewcenter.in/wp-content/uploads/2015/05/xreviewcenter.jpg 560w, https://www-reviewcenter-in.cdn.ampproject.org/ii/w820/s/www.reviewcenter.in/wp-content/uploads/2015/05/xreviewcenter.jpg 820w" src="https://www-reviewcenter-in.cdn.ampproject.org/i/s/www.reviewcenter.in/wp-content/uploads/2015/05/xreviewcenter.jpg" class="i-amphtml-fill-content i-amphtml-replaced-content" sizes="(max-width: 1920px) 205px, 100vw">`

**Origin**
<img width="795" alt="origin" src="https://user-images.githubusercontent.com/41326532/61642451-ce708d80-ac98-11e9-907b-1b78c2ecc322.png">

**AMP Cache**
<img width="798" alt="amp-cache" src="https://user-images.githubusercontent.com/41326532/61642516-ecd68900-ac98-11e9-81e9-bbc3d796748f.png">

- [Cached GIST](https://gist.github.com/jamesozzie/6e91ac0c46729ef2218d39661ad2cb96)
- [Origin GIST](https://gist.github.com/jamesozzie/84c95750b9788f6af4880dc66193f7bb)

## How do we reproduce the issue?

1. Install and activate the NewsPaper theme on a WordPress website
2. Add the themes mobile plugin (which renders it own's AMP templates)
3. Compare difference on AMP origin logo and AMP Cache logo

## What browsers are affected?

All browsers? Some specific browser? What device type?

## Which AMP version is affected?

Is this a new issue? Or was it always broken? Paste the version of AMP where you saw this issue. (You can find the version printed in your browser's console.)

**Comment by Gregable** [comment]
Here's what I see on the origin site, newlines added for readability:

```
<amp-img
  class="td-retina-data amp-wp-enforced-sizes"
  data-retina="https://www.reviewcenter.in/wp content/uploads/2015/05/xreviewcenter.jpg"
  src="https://www.reviewcenter.in/wp-content/uploads/2015/05/xreviewcenter.jpg" 
  alt="Review Center India" 
  title="Review Center India"
  width="250" height="38"
  layout="intrinsic">
```

And on the cache:

```
<amp-img 
  alt="Review Center India" 
  class="td-retina-data amp-wp-enforced-sizes" 
  data-retina=https://www.reviewcenter.in/wp-content/uploads/2015/05/xreviewcenter.jpg
  layout=intrinsic
  src=https://www-reviewcenter-in.cdn.ampproject.org/i/s/www.reviewcenter.in/wp-content/uploads/2015/05/xreviewcenter.jpg 
  srcset="https://www-reviewcenter-in.cdn.ampproject.org/ii/w270/s/www.reviewcenter.in/wp-content/uploads/2015/05/xreviewcenter.jpg 270w, https://www-reviewcenter-in.cdn.ampproject.org/ii/w560/s/www.reviewcenter.in/wp-content/uploads/2015/05/xreviewcenter.jpg 560w, https://www-reviewcenter-in.cdn.ampproject.org/ii/w820/s/www.reviewcenter.in/wp-content/uploads/2015/05/xreviewcenter.jpg 820w"
  title="Review Center India"
  width=250 height=38>
```

The only differences are the URLs and the srcset. The image URLs are the same size as well. There must be something interesting going on with CSS, but I'm not entirely sure what to look for.

**Comment by jridgewell** [comment]
I think it's a UI issue with `layout=intrinsic` and `srcset`.

**Comment by jridgewell** [comment]
No, it's the automatic `sizes` attribute.

**Comment by Gregable** [comment]
@jridgewell Is the fix something the AMP cache needs to implement?

**Comment by jridgewell** [comment]
No, it's purely client side.

**Comment by cathyxz** [comment]
I think layout `intrinsic` doesn't like it when you try to force override it to `position: absolute` (as you can see in an isolated vanilla repro here: https://codepen.io/cathyxz/pen/qeNOow?editors=1000). 

You can get it to go back to the correct size by removing the `position: absolute` on the `<amp-img>`, but that un-centers the image. You can recenter it via other means (e.g. https://codepen.io/cathyxz/pen/EqyNPN?editors=1000). I'm not 100% convinced that this is due to auto-sizes because you can disable `auto-sizes` by removing the `srcset` attribute (which I did here: https://codepen.io/cathyxz/pen/ZgebGz?editors=1000, but it still shows the incorrect size), or providing a `sizes` attribute yourself.  

This is probably not directly related, but WP has had issues with layout `intrinsic` before, and added a `object-fit: contain` hack to all layout intrinsic `<amp-img>`s per https://github.com/ampproject/amphtml/issues/21371. This might also cause unexpected behavior. 

If the goal here is just to center the image and have it grow to the max possible allowed size, I think using a centering strategy that does not involve `position: absolute` will fix it. I'll have to do a bit more digging to figure out why the cached version behaves differently than the non-cached.

**Comment by jridgewell** [comment]
The test page is flawed in some way, but I don't know what. Toggling `position: absolute` on the CDN page doesn't change the size of the image. Removing the `sizes` attribute does, though.

**Comment by cathyxz** [comment]
You're right. I'm missing this piece of CSS: 

```
.td-main-menu-logo amp-img img {
    width: auto;
    height: auto;
    min-width: 0;
    min-height: 0;
}
```
Okay, a few things are happening here. 
1. Layout Intrinsic doesn't play nice with `position: absolute`. 
2. We generate `srcset` on the cache, and `sizes` on the client, so it's possible to have smaller images than the original. 
3. There is a line of CSS here (width and height auto) forcing the `<img>` to display at its actual size (i.e. the size of the source image). 

Another observation, the CDN (problematic) version, will size the header image correctly when the viewport width is 820px or above. It will also size the header image correctly when the viewport width is 360px or below.  

The way `auto-sizes` works, is that it assumes that prior to `layoutCallback`, the `<amp-img>` has already been correctly sized. Therefore it measures the `<amp-img>`'s width in pixels, as well as the viewport width in pixels, and then sets a default entry, saying: for my current viewport size, my image should be the current size, then in `onMeasureChanged`, it recalculates this same logic and replaces the sizes attribute if the image has increased in width. For images that are not layout `fixed`, it calculates a viewport ratio. I'm noticing that this viewport ratio calculation is wrong. It's set to be `100vw`, when the header image in reality should be much smaller than `100vw`.

**Comment by cathyxz** [comment]
Layout Intrinsic: 

> With an intrinsic layout, the element takes the space available and resizes its height automatically to the aspect ratio given by the width and height attributes until it reaches the element's natural size or reaches a CSS constraint.

https://amp.dev/documentation/guides-and-tutorials/learn/amp-html-layout/layouts_demonstrated/

The definition of Layout Intrinsic, since it depends on the image's natural size, is inherently incompatible with autogenerated `srcset` and `sizes`, since what we do is generate smaller versions of the image, which prevents the image from growing to its intended full size. It's probably not necessary even to generate a `srcset` for layout `intrinsic`, since it's never intended to make use of that anyway. But since the cache already does it, what we can do for now is just opt layout intrinsic out of auto-sizes, so that it's unaffected.

**Comment by leafsy** [comment]
@cathyxz Thinking of picking this up first. Would it simply involve adding a check for intrinsic layout in the `maybeGenerateSizes` method of `amp-img.js`?

**Comment by cathyxz** [comment]
Yup! You can send a PR and tag me (@cathyxz) in the PR description for review. You should also add a unit test (in `test-amp-img.js`) to verify that sizes are are not generate for layout intrinsic.

**Comment by cathyxz** [comment]
So removing srcset on this bug actually still doesn't fix it either. But removing either the `min-height: 0` or `min-width: 0` on the `img` tag in the cached version does.

**Comment by cathyxz** [comment]
Okay I think I reduced this to a minimal repro: 
Cached version with srcset: https://codepen.io/cathyxz/pen/WNezBmv?editors=1000
Non-cached version without srcset: https://codepen.io/cathyxz/pen/gOYeNmb?editors=1000

I think this is just because the 
```
.td-main-menu-logo amp-img img {
    width: auto;
    height: auto;
    min-width: 0;
    min-height: 0;
}
```

styling forces the `img` tag to its natural size, which can be smaller for a lower browser width when the `srcset` attribute is generated, regardless of whether or not `sizes` is present. This repros in Chrome and Firefox. 

I also looked into the implementation with layout intrinsic. Our documentation is misleading in the sense that it references "the element's natural size". 

> With an intrinsic layout, the element takes the space available and resizes its height automatically to the aspect ratio given by the width and height attributes until it reaches the element's natural size or reaches a CSS constraint.

This is incorrect. It's actually not bound by the element's natural size, but by the size defined by the `width` and `height` attributes (you can toggle the width and height attributes and alter the natural size of the amp-img src in this sample to test: https://codepen.io/cathyxz/pen/eYOeQoY). 

So I think the real issue here isn't that layout intrinsic is incompatible with `auto-sizes`, but rather that this particular styling isn't compatible with sizes or srcset, since what it's trying to do is force the image to its natural size, which can change based on the srcsets that we generate: 

```
amp-img img {
    width: auto;
    height: auto;
    min-width: 0;
    min-height: 0;
}
``` 

I think our layout intrinsic documentation needs an update. But I think my recommendation here is just to remove this styling from the theme, and I'm not sure if there's anything else we should do in AMP to fix this.

**Comment by honeybadgerdontcare** [comment]
Thank you for the detailed explanation and taking the time to find the underlying issue. From what you've discovered there is nothing the cache can do to improve the situation for users.

**Comment by cathyxz** [comment]
Per https://github.com/ampproject/amphtml/issues/23453#issuecomment-530151343, since we've updated the documentation and added clarifying tests, this issue is safe to close out.
